source/client.c

```diff
@@ -580,42 +609,153 @@ static bool s_subscribe_send(uint16_t message_id, bool is_first_attempt, void *u
-            s_on_topic_clean_up,
-            task_arg)) {
-
-        goto handle_error;
-    }
-
     return false;
 
 handle_error:
 
+    /* Apparently you have to have an expression directly after a label, and defining a variable doesn't count. Neat. */
+    (void)0;
+
+    const size_t num_added_topics = aws_array_list_length(&subscribe.topic_filters);
+    for (size_t i = 0; i < num_added_topics; ++i) {
+        struct aws_mqtt_subscription *sub = NULL;
+        aws_array_list_get_at_ptr(&subscribe.topic_filters, (void **)&sub, i);
+
+        aws_mqtt_topic_tree_remove(&task_arg->connection->subscriptions, &sub->topic_filter);
+    }
+
     aws_mqtt_packet_subscribe_clean_up(&subscribe);
 
     if (message) {
         aws_channel_release_message_to_pool(task_arg->connection->slot->channel, message);
     }
 
-    aws_mqtt_topic_tree_remove(&task_arg->connection->subscriptions, &topic_cursor);
-
     aws_mem_release(task_arg->connection->allocator, task_arg);
 
     return true;
 }
 
+static void s_subscribe_complete(struct aws_mqtt_client_connection *connection, uint16_t packet_id, void *userdata) {
+
+    struct subscribe_task_arg *task_arg = userdata;
+
+    if (task_arg->on_suback) {
+        task_arg->on_suback(connection, packet_id, &task_arg->topics, task_arg->on_suback_ud);
+    }
+
+    aws_array_list_clean_up(&task_arg->topics);
+
+    aws_mem_release(task_arg->connection->allocator, task_arg);
+}
+
 uint16_t aws_mqtt_client_connection_subscribe(
+    struct aws_mqtt_client_connection *connection,
+    const struct aws_array_list *topic_filters,
+    aws_mqtt_suback_fn *on_suback,
+    void *on_suback_ud) {
+
+    assert(connection);
+
+    struct subscribe_task_arg *task_arg = aws_mem_acquire(connection->allocator, sizeof(struct subscribe_task_arg));
+    if (!task_arg) {
+        goto handle_error;
+    }
+
+    task_arg->connection = connection;
+    task_arg->on_suback = on_suback;
+    task_arg->on_suback_ud = on_suback_ud;
+
+    const size_t num_topics = aws_array_list_length(topic_filters);
+
+    aws_array_list_init_dynamic(&task_arg->topics, connection->allocator, num_topics, sizeof(void *));
+
+    for (size_t i = 0; i < num_topics; ++i) {
+
+        struct aws_mqtt_topic_subscription *request = NULL;
+        aws_array_list_get_at_ptr(topic_filters, (void **)&request, i);
+
+        if (!aws_mqtt_is_valid_topic_filter(&request->topic)) {
+            aws_raise_error(AWS_ERROR_MQTT_INVALID_TOPIC);
+            goto handle_error;
+        }
+
+        struct subscribe_task_topic *task_topic =
+            aws_mem_acquire(connection->allocator, sizeof(struct subscribe_task_topic));
+        task_topic->connection = connection;
+        task_topic->request = *request;
+
+        task_topic->filter = aws_string_new_from_array(
+            connection->allocator, task_topic->request.topic.ptr, task_topic->request.topic.len);
+        if (!task_topic->filter) {
+            goto handle_error;
```

Comment:
if this fails, `task_topic` will leak because the error handling logic iterates over `task_arg->topics` but we haven't pushed this `task_topic` into the list yet